sec/security.go

```diff
@@ -24,3 +24,17 @@ type SecureTransport interface {
 	// SecureOutbound secures an outbound connection.
 	SecureOutbound(ctx context.Context, insecure net.Conn, p peer.ID) (SecureConn, error)
 }
+
+// A SecureMuxer is a wrapper around SecureTransport which can select security protocols
+// and open outbound connections with simultaneous open.
+type SecureMuxer interface {
+	// SecureInbound secures an inbound connection.
+	// The returned boolean indicates whether the connection should be trated as a server
+	// connection; in the case of SecureInbound it should always be true.
+	SecureInbound(ctx context.Context, insecure net.Conn) (SecureConn, bool, error)
```

Comment:
Per our discussion, let's figure out _when_ this can happen. We normally won't get a "simultaneous accept".

Reply:
Well, there is symmetry in having the boolean indicator in the interface, plus it opens the door to active implementations that can deal with the situation.

Reply:
@vyzo I have the same concern as @Stebalien. Can you explain what you mean by active implementations? Is that some kind of hole punching technique you're referring to?

Reply:
I don't have anything that is doing active listen yet, but we could conceivably solve the accept-accept scenario in hole punching (that according to the literature _can_ happen, we'll see in practice).
Other than that, it's symmetry of the interface.

Reply:
What is the accept-accept scenario? Opening a TCP socket is a purely local operation that doesn't send any packets on the wire, isn't it? So this can't possibly lead to the establishment of a connection.

Reply:
It something that can happen with hole-punching, because you are both listening and opening a connection on the same time.
So it is conceivable that the `SYN`s will get through the NAT/firewall and the OS routes them to the listening socket instead of the connecting socket.
It is supposed to be unlikely, but something that can happen nonetheless.